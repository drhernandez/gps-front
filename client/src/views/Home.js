/*global google*/
import React from "react";
import {
  Container,
  Row,
  Col,
  Card,
  CardHeader,
  CardBody,
  FormSelect,
  Form,
} from "shards-react";
import PageTitle from "./../components/common/PageTitle";
import Gmap from "./../components/maps/Gmap";
import { Marker } from "./../components/maps/Marker";
import to from "await-to-js";
//services
import { UsersService, VehiclesService } from "../api/services"

export default class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      vehicles: [],
      trackings: [],
      markers: [],
      center: null,
      zoom: null
    }
  };

  async loadVehicles() {
    const [err, vehicles] = await to(UsersService.getVehiclesByUserID(10));
    if (!err && vehicles) {
      this.setState({
        vehicles: vehicles
      });
    }
    else {
      console.log(err);
    }
  }

  async getCurrentLocation(vehicleID) {
    const [err, location] = await to(VehiclesService.getCurrentLocation(vehicleID));
    if (!err && location) {
      const marker = new Marker(location.id, location.label, undefined, google.maps.Animation.DROP, location.lat, location.lng );
      this.setState({
        trackings: [location],
        markers: [marker],
        center: marker.position,
        zoom: 15
      })
    }
    else {
      console.log(err);
    }
  }

  handleDeviceOnChange(event) {
    event.persist();
    this.getCurrentLocation(event.target.value);
  }

  componentWillMount() {
    this.loadVehicles();
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4">
        {/* Page Header */}
        <Row noGutters className="page-header py-4">
          <PageTitle title="Ubicación actual" subtitle="Mapa" className="text-sm-left mb-3 col-sm-12" />
        </Row>

        <Row>
          <Col>
            <Card small className="mb-4">
              <CardHeader>
                <Form className="add-new-post">
                  <FormSelect id="feInputState" defaultValue="default" onChange={(event) => this.handleDeviceOnChange(event)}>
                    <option value="default" disabled>Elija un vehículo...</option>
                    {this.state.vehicles.map((vehicle, idx) => (
                      <option key={vehicle.device_id} value={vehicle.id}>{`${vehicle.type} - ${vehicle.plate}`}</option>
                    ))}
                  </FormSelect>
                </Form>
              </CardHeader>
              <CardBody>
                <Gmap
                  isMarkerShown
                  containerElement={<div style={{ height: '450px' }} />}
                  mapElement={<div style={{ height: `100%` }} />}
                  zoom={this.state.zoom}
                  markers={this.state.markers}
                  center={this.state.center}
                />
              </CardBody>
            </Card>
          </Col>
        </Row>
      </Container>
    );
  }
}