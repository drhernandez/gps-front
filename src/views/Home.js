import React from "react";
// import PropTypes from "prop-types";
import {
  Container,
  Row,
  Col,
  Card,
  CardHeader,
  CardBody,
  FormSelect,
  Form,
} from "shards-react";
import PageTitle from "./../components/common/PageTitle";
import Gmap from "./../components/maps/Gmap";
//services
import UsersService from "./../api/services/usersService";
import VehiclesService from "./../api/services/vehiclesService";

export default class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      vehicles: [],
      trackings: [],
      center: null,
      zoom: null
    }
  };

  async loadDevices() {
    const usersService = new UsersService();
    const vehicles = await usersService.getVehiclesByUserID(10);
    this.setState({ 
      vehicles: vehicles 
    });
  }

  async getCurrentLocation(vehicleID) {
    const vehiclesServices = new VehiclesService();
    const currentLocation = await vehiclesServices.getCurrentLocation(vehicleID);
    if (currentLocation !== null || currentLocation !== undefined) {
      this.setState({
        trackings: [currentLocation],
        center: currentLocation.position,
        zoom: 15
      })
    }
  }

  handleDeviceOnChange(event) {
    event.persist();
    this.getCurrentLocation(event.target.value);
  }

  componentWillMount() {
    this.loadDevices();
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4">
        {/* Page Header */}
        <Row noGutters className="page-header py-4">
          <PageTitle title="Where is my car?!" subtitle="Dashboard" className="text-sm-left mb-3 col-sm-12" />
        </Row>

        <Row>
          <Col>
            <Card small className="mb-4">
              <CardHeader>
                <Form className="add-new-post">
                  <FormSelect id="feInputState" defaultValue="default" onChange={(event) => this.handleDeviceOnChange(event)}>
                    <option value="default" disabled>Elija un veh√≠culo...</option>
                    {this.state.vehicles.map((vehicle, idx) => (
                      <option key={vehicle.device_id} value={vehicle.id}>{`${vehicle.type} - ${vehicle.plate}`}</option>
                    ))}
                  </FormSelect>
                </Form>
              </CardHeader>
              <CardBody>
                <Gmap
                  googleMapURL="http://maps.google.com/maps/api/js?key=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
                  loadingElement={<div style={{ height: `100%` }} />}
                  containerElement={<div style={{ height: '450px' }} />}
                  mapElement={<div style={{ height: `100%` }} />}
                  zoom={this.state.zoom}
                  markers={this.state.trackings}
                  center={this.state.center}
                />
              </CardBody>
            </Card>
          </Col>
        </Row>
      </Container>
    );
  }
}