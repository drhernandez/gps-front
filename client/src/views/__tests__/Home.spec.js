import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

jest.mock("../../api/services/vehiclesService", () => ({ searchVehicles: jest.fn(), getCurrentLocation: jest.fn() }));
import VehiclesService from '../../api/services/vehiclesService';

import Home from '../Home';

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

let store;
let searchMock;
let locationMock;

describe('Home component', () => {
  beforeEach(async () => {
    jest.resetModules();
    
    store = mockStore({
      userInfo: {
        id: 1
      },
    });

    searchMock = {
      data: [
        {
          id: 1,
          brand: "Frod",
          brand_line: "Fiesta",
          plate: "AA111AA"
        }
      ],
      paging: {
        total: 1
      }
    }

    locationMock = {
      "id": 1,
      "device_id": 1,
      "lat": -31.422108,
      "lng": -64.18643
    }

    await VehiclesService.searchVehicles.mockImplementation(() => {
      return Promise.resolve(searchMock);
    });
    await VehiclesService.getCurrentLocation.mockImplementation(() => {
      return Promise.resolve(locationMock);
    });
  });

  it("snapshot renders", () => {

    const component = renderer.create(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should exist the vehicle on FormSelect options", async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper.find('option').at(1).text()).toBe("Frod Fiesta - AA111AA");
  });

  it("should get location of selected vehicle", async () => {
    debugger
    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('CardBody').props().children.props.markers).toHaveLength(0);

    const mockedEvent = {
      target: {
        value: 1
      },
      preventDefault: () => { },
      persist: () => { }
    }
    wrapper.find('FormSelect').props().onChange(mockedEvent);

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('CardBody').props().children.props.markers).toHaveLength(1);
    expect(wrapper.find('CardBody').props().children.props.markers[0].id).toBe(locationMock.id);
    expect(wrapper.find('CardBody').props().children.props.markers[0].position.lat).toBe(locationMock.lat);
    expect(wrapper.find('CardBody').props().children.props.markers[0].position.lng).toBe(locationMock.lng);
  });
});