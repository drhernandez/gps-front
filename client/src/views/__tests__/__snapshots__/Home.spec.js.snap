// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component snapshot renders 1`] = `
<div
  className="main-content-container px-4 container-fluid"
>
  <div
    className="page-header py-4 no-gutters row"
  >
    <div
      className="mb-3 col-sm-12 text-center text-lg-left mb-sm-0 col-12 col-sm-4"
    >
      <span
        className="text-uppercase page-subtitle"
      >
        Mapa
      </span>
      <h3
        className="page-title"
      >
        Ubicación actual
      </h3>
    </div>
    <div
      className="col"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="mb-4 card card-small"
      >
        <div
          className="card-header"
        >
          <form
            className="add-new-post"
          >
            <select
              className="form-control custom-select"
              defaultValue="default"
              id="feInputState"
              onChange={[Function]}
            >
              <option
                disabled={true}
                value="default"
              >
                Elija un vehículo...
              </option>
            </select>
          </form>
        </div>
        <div
          className="card-body"
        >
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
